PIPS/pip-2.md

```diff
@@ -0,0 +1,115 @@
+
+Example fee deduction implementation for lock:
+```
+struct TxArgs {
+  bytes fromAssetHash;
+  bytes toAssetHash;
+  bytes toAddress;
+  uint256 amount;
+  uint256 feeAmount;
+  bytes feeReceiverAddr;
+}
+
+function lock(
+  address fromAssetHash,
+  uint64 toChainId,
+  bytes memory targetProxyHash,
+  bytes memory toAssetHash,
+  bytes memory toAddress,
+  uint256 amount,
+  bool deductFeeInLock,
+  uint256 feeAmount,
+  bytes memory feeReceiverAddr
+)
+  public
+{
+  bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+  require(registry[key] == true);
+
+  // Use SafeMath to ensure balances do not overflow
+  balances[key] = balances[key].add(amount);
+
+  // transfer tokens from user to LockProxy
+  require(_transferToContract(fromAssetHash, amount), "transfer asset from fromAddress to lock_proxy contract failed!");
+
+  TxArgs memory txArgs = TxArgs({
+      fromAssetHash: fromAssetHash,
+      toAssetHash: toAssetHash,
+      toAddress: toAddress,
+      amount: remainingAmount,
+      feeAmount: feeAmount,
+      feeReceiverAddr: feeReceiverAddr
+  });
+
+  uint256 remainingAmount = amount;
+  if (feeAmount > 0 && deductFeeInLock) {
+    // ensure that there is no overflow
+    remainingAmount = amount.sub(feeAmount);
+
+    // change fee amount to zero as the fee has already been deducted
+    txArgs.feeAmount = 0;
+  }
+
+  bytes memory txData = _serializeTxArgs(txArgs);
+
+  require(eccm.crossChain(toChainId, targetProxyHash, "unlock", txData), "EthCrossChainManager crossChain executed error!");
+
+  emit LockEvent(address(this), toChainId, targetProxyHash, txData);
+}
+```
+
+Example fee deduction implementation for unlock:
+```
+function unlock(bytes memory argsBs, bytes memory fromContractAddr, uint64 fromChainId) onlyManagerContract returns (bool) {
+  TxArgs memory args = _deserializTxArgs(argsBs);
+  bytes32 key = hash(args.toAssetHash, fromChainId, fromContractAddr, args.fromAssetHash);
+
+  require(registry[key] == true);
+  require(balances[key] >= args.amount);
+
+  // Use SafeMath to ensure balances do not overflow
+  balances[key] = balances[key].sub(args.amount);
+
+  uint256 remainingAmount = args.amount;
+  if (args.feeAmount > 0) {
+    // ensure that there is no overflow
+    remainingAmount = amount.sub(args.feeAmount);
+
+    // transfer feeAmount to feeReceiverAddr
+    require(_transferFromContract(fromAddress, feeReceiverAddr, feeAmount), "transfer asset from lock_proxy contract to toAddress failed!");
+  }
+
+  // send tokens to `toAddress`
+  require(_transferFromContract(args.toAssetHash, toAddress, remainingAmount), "transfer asset from lock_proxy contract to toAddress failed!");
+
+  emit UnlockEvent(fromContractAddr, fromChainId, args.toAddress, remainingAmount);
+
+  return true;
+}
+```
+
+If `deductFeeInLock` is set to `true`, it is possible for a relayer to only send the `lock` transaction, claim their fee, and then not send the `unlock` transaction to the target chain.
```

Comment:
Even if the relayer does not relay tx from fromChain to PolyChain, the relayer will receive his fee since the operation of transferring fee to fee receiver is triggered by the locker, who is the msg.sender or user.

Reply:
you're right, will change it